### ISSUE TITLE
ES7 property initializers disappear when combined with certain decorator functions

### ISSUE THREAD

**Issue by namuol** [gaearon/react-proxy#23]
(Note: I encountered this while using `react-transform-webpack-hmr`, but I believe the issue to be internal to `createProxy`.)

When using a `@decorator` function that copies the original component-class into a new one, `static` properties specified using ES7 initializers are unaccounted for.

For example, here's a slightly modified version of `App.js` from `react-transform-boilerplate`:

```js
import React, { Component } from 'react';

// This is TypeScript's implementation of Extend:
var __extends = function (d, b) {
  for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p];
  function __() { this.constructor = d; }
  __.prototype = b.prototype;
  d.prototype = new __();
};

function decorate (BaseComponent) {
  function DecoratedComponent (...args) {
    console.log("Running decorated constructor!");
    BaseComponent.apply(this, ...args)
  }

  __extends(DecoratedComponent, BaseComponent);

  return DecoratedComponent;
}

@decorate
class Number extends Component {
  static defaultProps = {
    value: 1,
  };

  render() {
    return <h1>{this.props.value} (defaultProps={'' + Number.defaultProps})</h1>;
  }
}

export class App extends Component {
  render() {
    return (
      <div>
        <Number />
        <Number value={42} />
      </div>
    );
  }
}
```

We expect the rendered output to look like this:
![](https://i.gyazo.com/982fbc99947b3dec28b76b666ac25647.png)

But instead we see this:
![](https://i.gyazo.com/525771cdf64b41659cd36621477fab71.png)

For some reason, `Counter.defaultProps` is not transferred in TypeScript's implementation of `__extends` when it is defined with ES7 generators.

It may have something to do with how `__extends` uses `hasOwnProperty`, but I'm not familiar enough with how `react-proxy` works under the hood to really take a stab at what might be going on, here.

- If I remove the `@decorator`, it works as expected.
- If I change `@decorator` to use ES6's `extends`, (i.e. `class DerivedComponent extends Component ...`), it works as expected.
- If I specify `defaultProps` *after* the decorator runs (i.e. `Counter.defaultProps = ...`), it works as expected.

I first encountered this problem when using [`react-free-style`](https://github.com/blakeembrey/react-free-style), which is written in TypeScript and provides a decorator function that extends components.

The natural workaround is to avoid static ES7 property initializers, but many of us are already using them, [as encouraged by React](https://facebook.github.io/react/blog/2015/01/27/react-v0.13.0-beta-1.html#es7-property-initializers).

I [forked `react-transform-boilerplate`](https://github.com/namuol/react-transform-boilerplate-es7-initializer-bug) to make it easier to reproduce the problem:

```shell
git clone https://github.com/namuol/react-transform-boilerplate-es7-initializer-bug
cd react-transform-boilerplate-es7-initializer-bug
npm install
npm start
open http://localhost:3000
```

**Comment by gaearon** [comment]
Can you please add a failing test to react-proxy?
You can use existing tests as template.

**Comment by namuol** [comment]
No problem -- I'll try to get something written this weekend.

**Comment by namuol** [comment]
I couldn't reproduce the problem in isolation -- it turns out to be caused by a combination of `react-proxy` and `babel-plugin-react-transform`.

----

In a nutshell:

Decorators execute in the order *opposite* that they appear, and `react-transform` pushes its decorators at the *end* of the decorators list, so they therefore execute *before* any pre-existing decorators.

This means that the proxied version of the class is actually what's passed to `@decorator`.

Since `react-proxy` [uses `__proto__` to copy static methods and properties](https://github.com/gaearon/react-proxy/blob/36ce47e8335501c6a7ca20d02b1c001dabf1b804/src/createClassProxy.js#L45), the `b.hasOwnProperty(p)` of TypeScript's `__extend` method fails, and none of the static properties/methods are copied over.

----

There are at least two potential fixes:

1. In `react-proxy`: Manually copy the static methods and properties directly onto the proxied class
2. In `babel-plugin-react-transform`: Insert the plugin's decorator at the *front* of the decorators list so that it executes *after* all pre-existing decorators

Either one of these will fix this particular issue, but I get the feeling both fixes should be applied; what do you think?

Thanks!

**Comment by gaearon** [comment]
>Manually copy the static methods and properties directly onto the proxied class

Can you try to do this (without breaking the tests)? Another option is to put getters on all static properties/methods and proxy them to the newest available version.

>Insert the plugin's decorator at the front of the decorators list so that it executes after all pre-existing decorators

That won't work. We *want* to wrap the original class because in 90% of use cases decorator is higher-order component—not a monkeypatching pseudo-inheritance thing like `__extend`.

**Comment by namuol** [comment]
> Can you try to do this (without breaking the tests)? Another option is to put getters on all static properties/methods and proxy them to the newest available version.

[Here's what I have so far](https://github.com/namuol/react-proxy/blob/preserve-hasOwnProperty-behavior/src/createClassProxy.js#L33L72).

I added some simple tests that compare the results of `getOwnPropertyNames` on the proxy and the original component class.

This implementation *almost* passes every test; I just need to fix [this last one](https://github.com/namuol/react-proxy/blob/preserve-hasOwnProperty-behavior/test/static-descriptor.js#L159L176).

I'll try to find a complete solution within the next few days; in the meantime, let me know if you think this is the right direction. Thanks!

**Comment by gaearon** [comment]
Direction seems right. :+1: 
Thank you!

**Comment by gaearon** [comment]
@namuol 

Can you please see if #29 fixes the issue for you? I was writing it offline so unfortunately I didn't use your code, but it seems to pass the test cases (and I added a bunch of new ones too).

**Comment by gaearon** [comment]
#29 seems to have fixed the issue in your project.

**Comment by namuol** [comment]
No worries about not using the changes I had -- my solution was incomplete
anyway. I'm busy right now but I'll make sure to test this out later
tonight; if the tests pass I'm pretty confident it should fix #23. Thanks!

**Comment by gaearon** [comment]
Great. Should be fixed in `react-proxy@1.1.1` and `react-transform-hmr@1.0.1`.
Thank you very much for reporting and taking the initial stab at it.
